src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -181,142 +187,184 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
-        long movingEndByte = partitionEndByteRange;
-        long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
-        boolean newLineFound = false;
-        while (!newLineFound) {
-          movingEndByte += MyriaConstants.BYTE_OVERLAP_PARALLEL_INGEST;
-          // Create a stream to look for the new line
-          InputStream trailingEndInputStream =
-              source.getInputStream(partitionEndByteRange, movingEndByte);
-          InputStreamReader startStreamReader = new InputStreamReader(trailingEndInputStream);
-          int dataChar = startStreamReader.read();
-          while (dataChar != -1) {
-            char currentChar = (char) dataChar;
-            if (currentChar == '\n' || currentChar == '\r') {
-              newLineFound = true;
-              // Re-initialize the parser with the last row only
-              InputStream beginningOfRecord =
-                  source.getInputStream(bytePositionAtBeginningOfRecord, partitionEndByteRange);
-              InputStream concatenateEndOfRecord =
-                  source.getInputStream(partitionEndByteRange + 1, movingEndByte);
-              partitionInputStream =
-                  new SequenceInputStream(beginningOfRecord, concatenateEndOfRecord);
-              parser =
-                  new CSVParser(
-                      new BufferedReader(new InputStreamReader(partitionInputStream)),
-                      CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
-                      bytePositionAtBeginningOfRecord,
-                      0);
-              iterator = parser.iterator();
-              onLastRow = true;
-              record = iterator.next();
+
+      if (record != null) {
+        /*
+         * Here, we read the first byte of the stream if this is not the first worker. This is simply an optimization.
+         * If the previous character is not a new line, we discard the line.
+         */
+        if (lineNumber - 1 == 0 && partitionStartByteRange != 0) {
+          char currentChar = (char) partitionInputStream.read();
+          if (currentChar != '\n' && currentChar != '\r') {
+            discardedRecord = true;
+            /*
+             * This handles closes the stream in corner cases in which a worker might only have a fragment of a tuple as
+             * its entire partition. Since the server makes sure to provide workers a partition with a minimum size,
+             * this case could only happen with very large rows.
+             */
+            if (onLastRow) {
+              parser.close();
               break;
             }
-            dataChar = startStreamReader.read();
+          } else {
+            /*
+             * This handles the case in which the splitting occurs in the middle of \r\n.
+             */
+            currentChar = (char) partitionInputStream.read();
+            if (currentChar == '\n') {
```

Comment:
This diff is outdated, currently, I don't read directly from the ```partitionInputStream``` anymore for this case. 

I reverted this optimization in my newest commit.